Update component “text-inputs” (#146)
Zywave/booster: src/_areas/design-system/components/text-inputs.md

---
title: Text inputs
subtitle: A single-line box that the user can enter text into.
api: https://cdn.zywave.com/@zywave/zui-input@next/dist/custom-elements.json
demo: https://cdn.zywave.com/@zywave/zui-input@next/demo/index.html
mainComponentName: zui-input
includedElements:
  - zui-input
---
## Usage

A form control allows the user to enter text into a UI. They typically appear in forms.

![Text input usage](/images/textinput_usage.svg)

- - -

## Anatomy

The anatomy of a text input.

![Text input anatomy](/images/textinput_anatomy.svg)

1. **Label:** Text to inform users as to what information is requested
2. **Text input:** The form field where the user can input information
3. **Placeholder text** (optional): Text that provides an example of what the user should enter into the text input

<br>

#### Sizing & spacing

The length of the input should be as close to the intended length of content as possible. This also applies to other form controls of a similar shape, like drop-down selects and text areas.

![Text Input sizing and spacing](/images/textinput_sizingandspacing.svg)

> See the [Design specs](https://xd.adobe.com/view/4bf410a7-cbe1-4642-9756-529baa10276d-1ea7/grid) for detailed sizing and spacing information

Reviewer: ```suggestion
<docs-note>See the [Design specs](https://xd.adobe.com/view/4bf410a7-cbe1-4642-9756-529baa10276d-1ea7/grid) for detailed sizing and spacing information.</docs-note>
```

- - -

## States

Text inputs can display the following states: unselected, hover, focus, and disabled.

![Text input states](/images/textinput_states.svg)

<br>

#### Read only

The input is not editable at this time. Read-only text fields should be used when there is no action the user can take on the page to enable the field. The benefit of using the read-only input field over a body of text is it still has a focused state, meaning the user can navigate to it using their keyboard.

> It may be frustrating or confusing to users when an input field is read-only. You may consider using a tool-tip that informs them where they may be able to go to change the information that appears in a read-only input.

Reviewer: ```suggestion
Read-only Text Inputs may cause frustration or confusion. You may consider using a [Tooltip](/design-system/components/tooltip) that informs them where they may be able to go to change the information that appears in the read-only Text Input.
```

- - -

## Types

#### Validation

> A success text Input is outlined in green. Success message may appear nearby.

Reviewer: Also consider being more specific with your description of the input. Somethin like "A success Text Input is outlined with a 1px Green 500 border." 
```suggestion
A success Text Input is outlined in green. A success message may appear below or next to the input.
```

> ![](/images/textinput_success.svg)

Reviewer: image is not full-width

<br>

> An error text Input is outlined in red. Error message with details on the error will appear nearby.

Reviewer: Same recommendation as above.

> ![](/images/textinput_error.svg)

Reviewer: Image is not full-width

> For more information on validation, [see forms documentation on validation](https://booster.zywave.dev/design-system/patterns/forms/#validation).

Reviewer: ```suggestion
<docs-note> For more information on validation, [see Forms documentation on validation](https://booster.zywave.dev/design-system/patterns/forms/#validation). </docs-note>
```

<br>

#### Text Area

> A Text Area is a subset of the text input specifically used for entering multiple sentences of information because of this, the input size is larger (width and length) than a standard input area. Text areas can also use a scroll bar and expander tab in the bottom right corner, if necessary.

Reviewer: ```suggestion
A Text Area is a subset of the Text Input specifically used for entering multiple lines of information. The input size is larger (width and length) than a standard Text Input. Text Areas can also include a scroll bar and expander tab in the bottom right corner, if necessary.
```

> ![Text input text area](/images/textinput_textarea.svg)

Reviewer: Image is not full-width

> We recommend capping the max height of a text area so it doesn't grow too large for the UI it's in, then use a scrollbar for viewing text not within the viewing area. 150px is typically enough to get a good sense of the text within, but not blow out your screen.

Reviewer: ```suggestion
We recommend capping the max-height of a Text Area so it doesn't use too much space within a UI. If it expands beyond the max-height, then a scroll bar can be used to view text that is not visible in the Text Area. We recommend starting with a max-height of 150px.  
```

> ![Text input text area scroll](/images/textinput_textarea_scroll.svg)

Reviewer: Image is not full-width

<br>

#### Text inputs with actions

> On rare occasions, you can put the button next to the input, which is often seen in search. While there's nothing especially wrong with putting the button below the input, putting it next to it saves space and looks cleaner. But, do not do this on standard forms that happen to have just 1 field. It's inconsistent and unconventional. The button should be labeled with text unless there is a very clear and deliberate icon that can be used.

Reviewer: ```suggestion
In rare occasions, a Button can be placed directly next to the Text Input, most commonly seen on search fields. The close proximity of this Button saves space in the UI and ties an action directly to a Text Input. The button should have a text label or use a very clear and deliberate icon. This button placement should be used rarely and should not be used for most single Text Inputs or on Forms. 
```

![Text Input actions](/images/textinput_actions.svg)

<br>

> For more information on Search, [see Search documentation](https://booster.zywave.dev/design-system/components/search/?tab=usage).

Reviewer: ```suggestion
For more information on Search, [see the guidelines](https://booster.zywave.dev/design-system/components/search/?tab=usage).
```

<br>

#### Input prefix

> Added to front of the input field to specify what should be entered, like a dollar sign. To add more attention a container can be added to the styling. These labels are typically an icon.

Reviewer: ```suggestion
An input prefix can be added to front of the Text Input to specify what should be entered, such as a dollar sign. To add more attention a container can be added to the styling. These labels are typically an icon.
```

Reviewer: I am not sure what the second sentence here means.

> ![Text Input input prefix](/images/textinput_input_prefix.svg)

Reviewer: image is not full-width

<br>

#### Input suffix

> Added to back of the input field to specify what should be entered, like a percentage symbol. These labels are typically an icon.

Reviewer: ```suggestion
An input suffix can be added to back of the Text Input to specify what should be entered, such as a percentage symbol. These labels are typically an icon.
```

Reviewer: Is the prefix also usually an icon? Do we even want to say this? In your example below the suffix is numbers.

> ![Text Input input suffix](/images/textinput_input_suffix.svg)

Reviewer: image is not full-width

<br>

#### Input prefix & suffix

> Added to front and back of the input field to specify what should be entered, like a dollar sign and a decimal number to represent a dollar amount.

Reviewer: ```suggestion
An input prefix and suffix can be added to front and back of the Text Input to specify what should be entered, such as a dollar sign and a decimal number to represent a dollar amount.
```

![Text Input input prefix and suffix](/images/textinput_input_prefixandsuffix.svg)

- - -

## Best practices

#### Labeling

<docs-grid columns="2">

<div>

![Text input labeling do](/images/textinput_labeling_do.svg)

<docs-do>

> Use clear and concise labeling. Label text is used to inform users as to what information is requested for a text input. 

Reviewer: ```suggestion
Use clear and concise labels to inform users as to what information they should enter in a Text Input.
```

</docs-do>

</div>

<div>

![Labeling don't](/images/textinput_labeling_dont.svg)

<docs-do-not>
Avoid using sentences or long phrases for label text. 
</docs-do-not>

</div>

</docs-grid>

<docs-spacer>

</docs-spacer>

> #### Place holder text

Reviewer: We had made a determination a while ago that this should be left blank in most situations to avoid redundancy. It is beneficial to use when there is specific formatting needed. You might want to rework this section.

<docs-grid columns="2">

<div>

![Text input placeholder text do](/images/textinput_placeholder_do.svg)

<docs-do>
Use placeholder text to inform users as to what information is requested for a dropdown. It provides an example of what you should enter in that field, and the format of that information when appropriate.
</docs-do>

</div>

<div>

![Text input placeholder text don't](/images/textinput_placeholder_dont.svg)

<docs-do-not>
Avoid repeating the same text used for the label as it is redundant information and does not inform the user of the format of information requested.
</docs-do-not>

</div>

</docs-grid>

<docs-spacer>

</docs-spacer>

#### Input length

<docs-grid columns="2">

<div>

![Text input placeholder text do](/images/textinput_size_do.svg)

<docs-do>

> Use an input size that is properly proportional to the expected user input

Reviewer: ```suggestion
Use an input width that is proportional to the expected user input
```

</docs-do>

</div>

<div>

![Text input placeholder text don't](/images/textinput_size_dont.svg)

<docs-do-not>

> Avoid using an input size that hides information or allows for excessive white space.

Reviewer: ```suggestion
Avoid using an input width that will cut-off information or allows for excessive white space.
```

</docs-do-not>

</div>

</docs-grid>

- - -

## Alternate considerations

> * If the user should be presented with supported or known values to choose from, consider one of the following:

Reviewer: ```suggestion
* If the user should be presented with a list of supported or known values to choose from, consider one of the following:
```

---
title: Text inputs
subtitle: A single-line box that the user can enter text into.
api: https://cdn.zywave.com/@zywave/zui-input@next/dist/custom-elements.json
demo: https://cdn.zywave.com/@zywave/zui-input@next/demo/index.html
mainComponentName: zui-input
includedElements:
  - zui-input
---
## Usage

> A form control allows the user to enter text into a UI. They typically appear in forms and dialogs.

Reviewer 2: ```suggestion
A form control allows the user to enter text into a UI. They typically appear in forms.
```
It's a little odd to include dialogs as a place they typically appear in. I think calling out forms is fine. Perhaps you have a specific reason to call out dialogs that I'm not thinking of?

![Text input usage](/images/textinput_usage.svg)

- - -

## Anatomy

The anatomy of a text input.

![Text input anatomy](/images/textinput_anatomy.svg)

1. **Label:** Text to inform users as to what information is requested
2. **Text input:** The form field where the user can input information
3. **Placeholder text** (optional): Text that provides an example of what the user should enter into the text input

<br>

#### Sizing & spacing

The length of the input should be as close to the intended length of content as possible. This also applies to other form controls of a similar shape, like drop-down selects and text areas.

![Text Input sizing and spacing](/images/textinput_sizingandspacing.svg)

See the [Design specs](https://xd.adobe.com/view/4bf410a7-cbe1-4642-9756-529baa10276d-1ea7/grid) for detailed sizing and spacing information

- - -

> ## States

Reviewer 2: Missing definitions for the different type of states.

> Text inputs can display the following states: unselected, hover, focus, entered, and disabled.

Reviewer 2: ```suggestion
Text inputs can display the following states: unselected, hover, focus, and disabled.
```
"Entered" isn't a valid state. I do see the value of having an example of what text would look like in a text input compared to placeholder text and disabled text. Maybe that should be part of another image example, like in the Usage section?

![Text input states](/images/textinput_states.svg)

- - -

## Types

#### Success

Text Input is outlined in green. Success message may appear nearby.

![](/images/textinput_success.svg)

<br>

#### Error

Text Input is outlined in red. Error message with details on the error will appear nearby.

> ![](/images/textinput_error.svg)

Reviewer 2: I would consider these validation types. Would it make sense to group them under that? I would also call out to the Forms pattern page for more details and other form info: https://booster.zywave.dev/design-system/patterns/forms/#validation

<br>

#### Read only

The input is not editable at this time. Read-only text fields should be used when there is no action the user can take on the page to enable the field. The benefit of using the read-only input field over a body of text is it still has a focused state, meaning the user can navigate to it using their keyboard.

> It may be frustrating or confusing to users when an input field is read-only. You may consider using a tool-tip that informs them where they may be able to go to change the information that appears in a read-only input.

Reviewer 2: This is considered a state for text inputs from a dev perspective. It would make more sense to move this up under the States section. You can update your States image example to rename the "Entered" state to "Read only" state.

![Text input read only](/images/textinput_readonly.svg)

<br>

#### Text Area

A Text Area is a subset of the text input specifically used for entering multiple sentences of information because of this, the input size is larger (width and length) than a standard input area. Text areas can also use a scroll bar and expander tab in the bottom right corner, if necessary.

![Text input text area](/images/textinput_textarea.svg)

We recommend capping the max height of a text area so it doesn't grow too large for the UI it's in, then use a scrollbar for viewing text not within the viewing area. 150px is typically enough to get a good sense of the text within, but not blow out your screen.

> ![Text input text area scroll](/images/textinput_textarea_scroll.svg)

Reviewer 2: Text areas are an entirely different component, so this info doesn't belong here. We can save this information for when there's a need to standardize and document text areas. From what I recall, we don't use text areas all that often. Do you agree?

<br>

> #### Text inputs with actions

Reviewer 2: Opportunity to link to our Search component here? https://booster.zywave.dev/design-system/components/search/

On rare occasions, you can put the button next to the input, which is often seen in search. While there's nothing especially wrong with putting the button below the input, putting it next to it saves space and looks cleaner. But, do not do this on standard forms that happen to have just 1 field. It's inconsistent and unconventional. The button should be labeled with text unless there is a very clear and deliberate icon that can be used.

![Text Input actions](/images/textinput_actions.svg)

<br>

> #### Input prefix

Reviewer 2: ```suggestion
#### Text input prefix
```

> Added to front of the input field to specify what should be entered, like a dollar sign. To add more attention a container can be added to the styling. These labels are typically an icon.

Reviewer 2: ```suggestion
Added to the beginning of the input field to specify what should be entered, such as a dollar sign. To draw more attention, a container can be added to the styling. These labels are typically an icon.
```

![Text Input input prefix](/images/textinput_input_prefix.svg)

<br>

> #### Input suffix

Reviewer 2: ```suggestion
#### Text input suffix
```

> Added to back of the input field to specify what should be entered, like a percentage symbol. These labels are typically an icon.

Reviewer 2: ```suggestion
Added to the end of the input field to specify what should be entered, like a percentage symbol. These labels are typically an icon.
```

![Text Input input suffix](/images/textinput_input_suffix.svg)

<br>

> #### Input prefix & suffix
> 
> Added to front and back of the input field to specify what should be entered, like a dollar sign and a decimal number to represent a dollar amount.
> 
> ![Text Input input prefix and suffix](/images/textinput_input_prefixandsuffix.svg)

Reviewer 2: ```suggestion
#### Text input prefix & suffix

A text input can have both a prefix and suffix to specify what should be entered, such as a dollar sign and a decimal number to represent a dollar amount.

![Text input prefix and suffix](/images/textinput_input_prefixandsuffix.svg)
```

- - -

## Behavior

#### Loading

> A loading icon appears on the right side of the input to indicate that something is happening.

Reviewer 2: Why should there be a loading icon in an input field? When should I use a loading icon in an input field?

Just wondering if we have a reason to document this, since this is not a feature in `<zui-input>`.

![Text input loading](/images/textinput_loading.svg)

- - -

## Best practices

#### Labeling

<docs-grid columns="2">

<div>

![Text input labeling do](/images/textinput_labeling_do.svg)

<docs-do>

> Use clear and concise labeling. Label text is used to inform users as to what information is requested for a text input. 

Reviewer 2: ```suggestion
Use clear and concise labels to inform users of what information is being requested. 
```

</docs-do>

</div>

<div>

![Labeling don't](/images/textinput_labeling_dont.svg)

<docs-do-not>

> Avoid using sentences or long phrases for label text. 

Reviewer 2: ```suggestion
Avoid using sentences or long phrases for labels. 
```

</docs-do-not>

</div>

</docs-grid>

<docs-spacer>

</docs-spacer>

> #### Place holder text

Reviewer 2: ```suggestion
#### Placeholder text
```

Reviewer 2: Thoughts on including a Do Not example of an input field with only the placeholder text and no label? It is bad practice to only rely on the placeholder text as the label since it disappears after you enter a value into the text input. https://www.nngroup.com/articles/form-design-placeholders/

<docs-grid columns="2">

<div>

![Text input placeholder text do](/images/textinput_placeholder_do.svg)

<docs-do>

> Use placeholder text to inform users as to what information is requested for a dropdown. It provides an example of what you should enter in that field, and the format of that information when appropriate.

Reviewer 2: ```suggestion
Supplement labels with placeholder text to further inform users of what information is being requested. For example, providing a telephone number format as placeholder text.
```

</docs-do>

</div>

<div>

![Text input placeholder text don't](/images/textinput_placeholder_dont.svg)

<docs-do-not>

> Avoid repeating the same text used for the label as it is redundant information and does not inform the user of the format of information requested.

Reviewer 2: ```suggestion
Avoid repeating the label as placeholder text since it is redundant information and does not inform the user of the format of information requested.
```

</docs-do-not>

</div>

</docs-grid>

<docs-spacer>

</docs-spacer>

> #### Input length

Reviewer 2: ```suggestion
#### Text input length
```

<docs-grid columns="2">

<div>

![Text input placeholder text do](/images/textinput_size_do.svg)

<docs-do>

> Use an input size that is properly proportional to the expected user input

Reviewer 2: ```suggestion
Set the text input length to the proper size of the expected user input.
```

</docs-do>

</div>

<div>

![Text input placeholder text don't](/images/textinput_size_dont.svg)

<docs-do-not>

> Avoid using an input size that hides information or allows for excessive white space.

Reviewer 2: ```suggestion
Avoid using a text input length that hides information or allows for excessive white space.
```
